warp/aggregator/aggregator.go

```diff
@@ -36,16 +70,134 @@ func (a *Aggregator) AggregateSignatures(ctx context.Context, unsignedMessage *a
+	if err != nil {
+		return nil, fmt.Errorf("failed to get validator set: %w", err)
+	}
+	if len(validators) == 0 {
+		return nil, fmt.Errorf("%w (SubnetID: %s, Height: %d)", errNoValidators, a.subnetID, pChainHeight)
+	}
+
+	return aggregateSignatures(ctx, a.client, unsignedMessage, validators, quorumNum, totalWeight)
+}
+
+func aggregateSignatures(
+	ctx context.Context,
+	client SignatureGetter,
+	unsignedMessage *avalancheWarp.UnsignedMessage,
+	validators []*avalancheWarp.Validator,
+	quorumNum uint64,
+	totalWeight uint64,
+) (*AggregateSignatureResult, error) {
+	var (
+		// [signatureLock] must be held when accessing [blsSignatures],
+		// [signersBitset], or [signatureWeight] in the goroutine below.
+		signatureLock   = sync.Mutex{}
+		signatures      = make([]*bls.Signature, 0, len(validators))
+		signersBitset   = set.NewBits()
+		signatureWeight = uint64(0)
+	)
+
+	// Create a child context to cancel signature fetching if we reach [maxNeededQuorumNum] threshold
+	signatureFetchCtx, signatureFetchCancel := context.WithCancel(ctx)
+	defer signatureFetchCancel()
+
+	wg := sync.WaitGroup{}
+	wg.Add(len(validators))
+	for i, validator := range validators {
+		var (
+			i         = i
+			validator = validator
+			// TODO: update from a single nodeID to the original slice and use extra nodeIDs as backup.
+			nodeID = validator.NodeIDs[0]
+		)
+		go func() {
+			defer wg.Done()
+
+			log.Debug("Fetching warp signature",
+				"nodeID", nodeID,
+				"index", i,
+			)
+
+			signature, err := client.GetSignature(signatureFetchCtx, nodeID, unsignedMessage)
+			if err != nil {
+				log.Debug("Failed to fetch warp signature",
+					"nodeID", nodeID,
+					"index", i,
+					"err", err,
+				)
+				return
+			}
+			log.Debug("Retrieved warp signature",
+				"nodeID", nodeID,
+				"index", i,
+				"signature", hexutil.Bytes(bls.SignatureToBytes(signature)),
+			)
+
+			if !bls.Verify(validator.PublicKey, signature, unsignedMessage.Bytes()) {
+				log.Debug("Failed to verify warp signature",
+					"nodeID", nodeID,
+					"index", i,
+					"signature", hexutil.Bytes(bls.SignatureToBytes(signature)),
+					"msgID", unsignedMessage.ID(),
+				)
+				return
+			}
+
+			// Add the signature and check if we've reached the requested threshold
+			signatureLock.Lock()
+			defer signatureLock.Unlock()
+
+			signatures = append(signatures, signature)
+			signersBitset.Add(i)
+			signatureWeight += validator.Weight
+			log.Debug("Updated weight",
+				"totalWeight", signatureWeight,
+				"addedWeight", validator.Weight,
+			)
+
+			// If the signature weight meets the requested threshold, cancel signature fetching
+			if err := avalancheWarp.VerifyWeight(signatureWeight, totalWeight, quorumNum, params.WarpQuorumDenominator); err == nil {
+				log.Debug("Verify weight passed, exiting aggregation early",
+					"maxNeededQuorumNum", quorumNum,
+					"totalWeight", totalWeight,
+					"signatureWeight", signatureWeight,
+				)
+				signatureFetchCancel()
+			}
```

Comment:
It seems that we double check this case once 
1- a go routine verifies this and cancels signature fetching. 
2- then the main go routine will again check this.

This also probably will try to fetch and verify signatures more than it actually requires: Sub go routines will be blocked at `signatureLock` for most of the active validators, meaning that they will perform this verification even if some other routine already verified before (and keep appending to already verified signature set). I wonder if splitting fetch + verification into two different routines would help this to not perform duplicate verifications. I.e
1- Spawn go routines for each validator and fetch signature
2- Fetched signatures can be fed to another channel 
3- Another goroutine could perform verification once it receives signature, without blocking fetching goroutines
4- When verification goroutine verifies the signature set it can cancel validator routines and returns with verified signatures.
5- Verification goroutine should wait until waitgroup is completed. If no validator goroutine left, return err. 

If verification goroutine successfully verifies, we wouldn't need to verify it again before aggregation.   

Alternatively we can try to exit from validator go routines after `signatureLock` is released (i.e with a check if `signatureFetchCancel` is called)